# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Прежде чем непосредственно отвечать на поставленный вопрос, необходимо понять, какие уникальные значения присутствуют в столбце 'debt'. 

Для этого применим к этому столбцу метод unique():

In [22]:
data['debt'].unique()

array([0, 1])

Как можно видеть, в столбце 'debt', означающем, имел ли клиент задолженность по возврату кредитов, имеется всего два значения: 0 и 1.

В таком случае, логично предположить, что 0 - клиент не имел задолженность, а 1 - клиент имел задолженность, поскольку 0 обычно означает отсутствие чего-либо (False), а 1 - присутствие (True).

Теперь перейдём непосредственно к ответу на вопрос.

Создадим сводную таблицу debt_child_pivot, в которой будет три столбца: 'Всего кредитополучателей', 'Всего должников', 'Доля должников', отсортированных по стол,цу 'children' (в порядке возрастания).
Для этого вызовем метод pivot_table(), где в качестве index возьмём столбец 'children', values - из столбца 'debt' и применим методы count(), sum() и mean().

In [23]:
debt_child_pivot = data.pivot_table(index='children', values='debt', aggfunc=['count', 'sum', 'mean'])

debt_child_pivot = debt_child_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников'})

debt_child_pivot

Всего кредитополучателей Всего должников Доля должников
                             debt            debt           debt
children                                                        
0                           14091            1063       0.075438
1                            4808             444       0.092346
2                            2052             194       0.094542
3                             330              27       0.081818
4                              41               4       0.097561
5                               9               0       0.000000

Чтобы можно было корректно сравнить разные по объёму категории клиентов в зависимости от количества детей между собой, примем максимальную долю за 100 %, а остальные будет считать относительно неё. Для этого в сводной таблице debt_child_pivot создадим столбец 'Сравнение долей должников':

In [24]:
debt_child_pivot['Сравнение долей должников'] = debt_child_pivot[('Доля должников', 'debt')] / debt_child_pivot[('Доля должников', 'debt')].max()

debt_child_pivot

Всего кредитополучателей Всего должников Доля должников  \
                             debt            debt           debt   
children                                                           
0                           14091            1063       0.075438   
1                            4808             444       0.092346   
2                            2052             194       0.094542   
3                             330              27       0.081818   
4                              41               4       0.097561   
5                               9               0       0.000000   

         Сравнение долей должников  
                                    
children                            
0                         0.773242  
1                         0.946547  
2                         0.969055  
3                         0.838636  
4                         1.000000  
5                         0.000000

**Вывод:** на основании полученных данных, можно сказать, что между количество детей у клиентов и возвратом кредита в срок существует зависимость. 

Доля должников среди клиентов, у которых отсутствуют дети, значительно ниже (более чем на 16 %) по сравнению с клиентами, имеющими от 1 до 4 детей, при этом с ростом количества детей (от 1 до 4 включительно) доля должников в целом имеет тенденцию к росту, за исключением клиентов с тремя детьми. Это, вероятно, объясняется тем, что бездетные клиенты несут наименьшую финансовую нагрузку, а клиенты с тремя детьми, с одной стороны, более ответственны и обеспечены большими мерами государственной поддержки по сравнению с теми, кто имеет 1-2 детей, а с другой стороны, имеют меньшую финансовую нагрузку относительно клиентов с четырьмя детьми. 

При этом у клиентов с пятью детьми должники отсутствуют, что наиболее вероятно объясняется нерепрезентативностью выборкой (численность = 9 человек) для данной категории. Однако не исключается вероятность того, что такие клиенты наиболее ответственны и не допускают просрочек по кредитным платежам.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Найдём сначала уникальные значения столбца 'family_status'.

Для этого применим к нему метод unique():

In [25]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

В столбце 'family_status' присутствует пять категорий, которые характеризуются наличием ('женат / замужем' и 'гражданский брак') 
и отсутствием ('вдовец / вдова', 'в разводе', 'Не женат / не замужем') партнёра.

Для ответа на поставленный вопрос создадим сводную таблицу debt_fam_stat_pivot, по аналогии с таблицей debt_child_pivot.

Для этого вызовем метод pivot_table(), где в качестве index возьмём столбец 'family_status', values - из столбца 'debt' и применим методы count(), sum() и mean(). Выведем полученную сводную таблицу в порядке возрастания значений в столбце 'Доля должников'.

In [26]:
debt_fam_stat_pivot=data.pivot_table(index='family_status', values='debt', aggfunc=['count', 'sum', 'mean'])

debt_fam_stat_pivot=debt_fam_stat_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников'})

debt_fam_stat_pivot.sort_values(by=('Доля должников', 'debt'))

Всего кредитополучателей Всего должников Доля должников
                                          debt            debt           debt
family_status                                                                
вдовец / вдова                             951              63       0.066246
в разводе                                 1189              84       0.070648
женат / замужем                          12261             927       0.075606
гражданский брак                          4134             385       0.093130
Не женат / не замужем                     2796             273       0.097639

Чтобы можно было корректно сравнить разные по объёму категории клиентов в зависимости от семейного положения между собой, примем максимальную долю за 100 %, а остальные будет считать относительно неё. Для этого в сводной таблице debt_fam_stat_pivot создадим столбец 'Сравнение долей должников':

In [27]:
debt_fam_stat_pivot['Сравнение долей должников'] = debt_fam_stat_pivot[('Доля должников', 'debt')] / debt_fam_stat_pivot[('Доля должников', 'debt')].max()

debt_fam_stat_pivot.sort_values(by=('Доля должников', 'debt'))

Всего кредитополучателей Всего должников Доля должников  \
                                          debt            debt           debt   
family_status                                                                   
вдовец / вдова                             951              63       0.066246   
в разводе                                 1189              84       0.070648   
женат / замужем                          12261             927       0.075606   
гражданский брак                          4134             385       0.093130   
Не женат / не замужем                     2796             273       0.097639   

                      Сравнение долей должников  
                                                 
family_status                                    
вдовец / вдова                         0.678476  
в разводе                              0.723556  
женат / замужем                        0.774334  
гражданский брак                       0.953816  
Не женат / не замужем                  1.000000

**Вывод:** между семейным положением и возвратом кредита в срок в целом есть зависимость.

Наименьшие доли должников (менее 75 % относительно максимально высокой доли должников) отмечаются среди клиентов, которые ранее состояли в браке: 'вдовец / вдова', 'в разводе', что вероятно обусловлено большей единоличной ответственностью таких клиентов и расчётом исключительно на собственные финансовые возможности.

Медианное положение занимают клиенты со статусом 'женат / замужем', что может объясняться, с одной стороны, довольно высокой финансовой нагрузкой по содержанию семьи (в том числе детей), а с другой стороны, финансовой поддержкой со стороны партнёрши / партнёра. 

Наиболее высокие доли должников (более 95% относительно максимально высокой доли должников) отмечаются для клиентов в гражданском браке и не женатых / не замужних. Такие показатели могут возникать по разным причинам, в частности, из-за большей финансовой свободы, необходимости содержания иждивенцев и др.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Ранее клиенты уже были разделены на пять категорий (A,B,C,D и E) в зависимости от уровня дохода.

Для ответа на поставленный вопрос создадим сводную таблицу debt_income_pivot, по аналогии с таблицей debt_child_pivot.

Для этого вызовем метод pivot_table(), где в качестве index возьмём столбец 'total_income_category', values - из столбца 'debt' и применим методы count(), sum() и mean(). Выведем полученную сводную таблицу в порядке возрастания значений в столбце 'Доля должников'.

In [28]:
debt_income_pivot=data.pivot_table(index='total_income_category', values='debt', aggfunc=['count', 'sum', 'mean'])

debt_income_pivot=debt_income_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников'})

debt_income_pivot.sort_values(by=('Доля должников', 'debt'))


Всего кредитополучателей Всего должников Доля должников
                                          debt            debt           debt
total_income_category                                                        
D                                          349              21       0.060172
B                                         5014             354       0.070602
A                                           25               2       0.080000
C                                        15921            1353       0.084982
E                                           22               2       0.090909

Чтобы можно было корректно сравнить разные по объёму категории клиентов в зависимости от уровня доходов между собой, примем максимальную долю за 100 %, а остальные будет считать относительно неё. Для этого в сводной таблице debt_income_pivot создадим столбец 'Сравнение долей должников':

In [29]:
debt_income_pivot['Сравнение долей должников'] = debt_income_pivot[('Доля должников', 'debt')] / debt_income_pivot[('Доля должников', 'debt')].max()

debt_income_pivot.sort_values(by=('Доля должников', 'debt'))

Всего кредитополучателей Всего должников Доля должников  \
                                          debt            debt           debt   
total_income_category                                                           
D                                          349              21       0.060172   
B                                         5014             354       0.070602   
A                                           25               2       0.080000   
C                                        15921            1353       0.084982   
E                                           22               2       0.090909   

                      Сравнение долей должников  
                                                 
total_income_category                            
D                                      0.661891  
B                                      0.776625  
A                                      0.880000  
C                                      0.934803  
E                                      1.000000

**Вывод:** между уровнем дохода клиентов и возвратом кредита в срок есть определённая зависимость. 

Так, с одной стороны, вполне логично, что категория E, т.е. клиенты с самым низким уровнем дохода (до 30000) имеет наиболее высокие показатели должников, так как такие клиенты обладают крайне небольшим объёмом свободных средств. Также  вполне объяснимо, что наиболее обеспеченные клиенты, относящиеся к категории A (уровень доходов - более 1000000), занимают медианное положение, поскольку это может быть обусловлено большой финансовой возможностью погасить задолженность по кредиту. Однако с другой стороны, выборки для категорий A и E являются малорепрезентативными, а потому в определённой степени их корректный анализ представляется затруднительным.

Кроме того, высокие показатели по доли должников отмечаются для клиентов категории C, в которой присутствует наибольшее количество человек (около 16 тысяч), имеющих доход от 50001 до 200000. Это, вероятно, свидетельствует о наличии у таких людей высокой финансовой нагрузки, обусловленной необходимостью содержания семьи и/или детей, выплаты ипотек и/или кредитов и в силу других причин.

Самые низкие доли должников наблюдаются для категорий D и B (около 66% и 78% относительно максимально высокой доли должников). Несмотря на значительно различающийся уровень доходов для этих категорий: D - от 30001 до 50000, B - от 200001 до 1000000, такие результаты могут быть связаны с большей финансовой ответственностью таких клиентов: с одной стороны относительно высокий доход и относительная финансовая свобода, а с другой - невозможность выплаты относительно больших задолженностей (по сравнению с уровнем дохода) по кредитам.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Ранее клиенты уже были разделены на четыре категории ('операции с автомобилем', 'операции с недвижимостью', 'проведение свадьбы' и 'получение образования') в зависимости от целей, на которые выдавался кредит.

Для ответа на поставленный вопрос создадим сводную таблицу debt_purpose_pivot, по аналогии с таблицей debt_child_pivot.

Для этого вызовем метод pivot_table(), где в качестве index возьмём столбец 'purpose_category', values - из столбца 'debt' и применим методы count(), sum() и mean(). Выведем полученную сводную таблицу в порядке возрастания значений в столбце 'Доля должников'.

In [30]:
debt_purpose_pivot=data.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'sum', 'mean'])

debt_purpose_pivot=debt_purpose_pivot.rename(columns={'count': 'Всего кредитополучателей', 'sum': 'Всего должников', 'mean': 'Доля должников'})

debt_purpose_pivot.sort_values(by=('Доля должников', 'debt'))


Всего кредитополучателей Всего должников  \
                                             debt            debt   
purpose_category                                                    
операции с недвижимостью                    10751             780   
проведение свадьбы                           2313             183   
получение образования                        3988             369   
операции с автомобилем                       4279             400   

                         Доля должников  
                                   debt  
purpose_category                         
операции с недвижимостью       0.072551  
проведение свадьбы             0.079118  
получение образования          0.092528  
операции с автомобилем         0.093480

Чтобы можно было корректно сравнить разные по объёму категории клиентов в зависимости от целей взятия кредита между собой, примем максимальную долю за 100 %, а остальные будет считать относительно неё. Для этого в сводной таблице debt_purpose_pivot создадим столбец 'Сравнение долей должников':

In [31]:
debt_purpose_pivot['Сравнение долей должников'] = debt_purpose_pivot[('Доля должников', 'debt')] / debt_purpose_pivot[('Доля должников', 'debt')].max()

debt_purpose_pivot.sort_values(by=('Доля должников', 'debt'))

Всего кредитополучателей Всего должников  \
                                             debt            debt   
purpose_category                                                    
операции с недвижимостью                    10751             780   
проведение свадьбы                           2313             183   
получение образования                        3988             369   
операции с автомобилем                       4279             400   

                         Доля должников Сравнение долей должников  
                                   debt                            
purpose_category                                                   
операции с недвижимостью       0.072551                  0.776119  
проведение свадьбы             0.079118                  0.846365  
получение образования          0.092528                  0.989814  
операции с автомобилем         0.093480                  1.000000

**Вывод:** 
наименьшие доли должников (менее 85% относительно максимально высокой доли должников) отмечены для тех клиентов, которые брали кредиты на операции с недвижимостью (в частности, на покупку и ремонт жилья) и создание семьи (организацию свадьбы). Предположительно, такие клиенты достаточно ответственно относятся к выплате кредитов в срок по вышеприведённым долгосрочным целям вследствие большого тела кредита (особенно в случае ипотеки) и длительных сроков выплаты кредита. 

В то же время наибольшие доли должников (более 98% относительно максимально высокой доли должников) наблюдаются для клиентов, взявших кредиты на получение образования и операции с автомобилем. Размеры и сроки выплаты кредитов, взятых для данных целей, как правило, достаточно невысокие, что, вероятно, может сказаться на финансовой ответственности клиентов и привести к появлению задолженностей. при погашении таких кредитов в срок.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
К возможным причинам появления пропусков в исходных данных можно отнести:
1. Необязательность заполнения всех необходимых полей при сборе исходных данных.
2. Частичная потеря данных при манипуляции с ними(например, при выгрузке) для последующего анализа.
3. Умышленное удаление части данных тем, кто имеет к ним доступ.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* количественные переменные могут сильно варьироваться, в них могут преобладать большие или малые значения, что приводит к некорректным (т.е. завышенным или заниженным) средним значениям, и для заполнения пропусков для количественных переменных лучше использовать медианное значение, которое вне зависимости от разброса данных корректно опишет выборку и поделит её пополам.

### Шаг 4: общий вывод.

Таким образом, для анализа были получены входные данные от кредитного отдела банка, а именно статистика о платёжеспособности клиентов. Необходимо было выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

В таблице с входными данными представлена информация о 21525 клиентах, распределённая по  следующим 12 столбцам:
- children — количество детей в семье;
- days_employed — общий трудовой стаж в днях;
- dob_years — возраст клиента в годах;
- education — уровень образования клиента;
- education_id — идентификатор уровня образования;
- family_status — семейное положение;
- family_status_id — идентификатор семейного положения;
- gender — пол клиента;
- income_type — тип занятости;
- debt — имел ли задолженность по возврату кредитов;
- total_income — ежемесячный доход;
- purpose — цель получения кредита.

Предобработка данных для анализа происходила следующим образом. Во входных данных присутствовали 2174 пропуска в столбцах 'days_employed' и 'total_income', которые были заполнены медианными значениями. 

Кроме того, в столбцах 'days_employed' и 'children' имелись аномальные значения (в частности, отрицательные числа), которые были либо заменены (в случае 'days_employed' вставлялись значения по модулю), либо удалены (в случае 'children'). 

В столбце 'education' присутствовали неявные дубликаты: одни и те же значения были записаны в разных регистрах, - которые были приведены к нижнему регистру. Явные дубликаты в виде 71 строчки были удалены. 

Данные в столбце 'total_income' были заменены с вещественного типа на целочисленный.

Таким образом, рекомендуется более тщательный подход (единообразное заполнение данных, исключение отрицательных значений и т.д.) при сборе данных во избежание появления во входных данных подобных пропусков, аномальных значений и дубликатов.

Категоризация данных осуществлялась для двух столбцов. Так, на основании значений в столбце 'total_income' был создан столбец 'total_income_category', в котором клиенты банка были разбиты на следующие категории:
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

Кроме того, был создан столбец 'purpose_category', в который вошли значения столбца 'purpose', сгруппированные так:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

В ходе исследовательского анализа необходимо было ответить на четыре поставленных вопроса:
- Есть ли зависимость между количеством детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

В результате ответов на эти вопросы можно заключить следующее:
- Добросовестный заёмщик не имеет детей, ранее состоял в браке, имеет доход от 30001 до 50000 и берёт кредит для операций с недвижимостью.
- Недобросовестный заёмщик имеет 4 детей, не женат / не замужем, имеет доход до 30000 и намерен взять кредит на операции с автомобилем.